### PR TITLE
Add naive concurrency limiting for drawable async load

### PR DESCRIPTION
**Pull request by peppy**
Will probably require more control over the scheduler being used, but this should be a sane starting point. PRing for further testing with osu!.

### CODE REVIEW

**Comment by swoolcock** [comment]
Will this help resolve concurrency issues in #1613 ?

**Comment by peppy** [comment]
Can you explain the concurrency issue there? I can't see anything on first read.

**Comment by swoolcock** [comment]
I mentioned the caveat that spamming updates could cause the `DelayedLoadWrapper` to be added even though there's a check to prevent it being added.  Would be great if it could store/use a cancellation token when the Drawable gets updated.

**Comment by peppy** [comment]
That looks like an issue with your logic. This will not fix it.

File: osu.Framework/Graphics/Containers/CompositeDrawable.cs [hunk 1]
```diff
@@ -102,6 +108,8 @@ protected override void LoadComplete()
 
         protected override void Dispose(bool isDisposing)
         {
+            cancelSource?.Cancel();
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Containers/CompositeDrawable.cs]: CTS must be disposed.

File: osu.Framework/Graphics/Drawable.cs [hunk 2]
```diff
@@ -81,8 +81,15 @@ private void dispose(bool isDisposing)
...
                 return;
 
-            //we can't dispose if we are mid-load, else our children may get in a bad state.
-            loadTask?.Wait();
+            try
+            {
+                //we can't dispose if we are mid-load, else our children may get in a bad state.
+                loadTask?.Wait();
+            }
+            catch
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Drawable.cs]: `catch (OperationCanceledException) { }` instead.

File: osu.Framework/IO/Network/WebRequest.cs [hunk 3]
```diff
@@ -344,7 +348,8 @@ private void internalPerform()
                 }
                 catch (Exception) when (timeoutToken.IsCancellationRequested)
                 {
-                    Complete(new WebException($"Request to {Url} timed out after {timeSinceLastAction / 1000} seconds idle (read {responseBytesRead} bytes, retried {RetryCount} times).", WebExceptionStatus.Timeout));
+                    Complete(new WebException($"Request to {Url} timed out after {timeSinceLastAction / 1000} seconds idle (read {responseBytesRead} bytes, retried {RetryCount} times).",
+                        WebExceptionStatus.Timeout));
```
Comment (smoogipoo) [inline osu.Framework/IO/Network/WebRequest.cs]: ?

File: osu.Framework/IO/Network/WebRequest.cs [hunk 4]
```diff
@@ -226,7 +229,8 @@ public byte[] ResponseData
                 throw new InvalidOperationException($"The {nameof(WebRequest)} has already been run.");
             try
             {
-                await Task.Factory.StartNew(internalPerform, TaskCreationOptions.LongRunning);
+                abortToken = new CancellationTokenSource(); // early initialisation so we can use it on the task itself.
```
Comment (smoogipoo) [inline osu.Framework/IO/Network/WebRequest.cs]: This is not dispose-safe. Potentially you want the following pattern:

```
using (abortToken = new CancellationTokenSource())
    await task_factory ...

/* internalPerform() */

using (abortToken ?? (abortToken = new CancellationToken())
    ...
```

File: osu.Framework/Platform/GameHost.cs [hunk 5]
```diff
@@ -494,6 +496,7 @@ private void bootstrapSceneGraph(Game game)
 
             Dependencies.Cache(root);
             Dependencies.CacheAs(game);
+            Dependencies.CacheAs(async_schedulers.ConcurrentScheduler);
```
Comment (smoogipoo) [inline osu.Framework/Platform/GameHost.cs]: `.Cache()`

File: osu.Framework/Testing/TestBrowser.cs [hunk 6]
```diff
@@ -382,15 +382,14 @@ public void LoadTest(Type testType = null, Action onCompletion = null, bool isDy
...
+            if (lastTest?.Parent != null)
             {
-                if (lastTest?.Parent != null)
-                {
-                    testContentContainer.Remove(lastTest.Parent);
-                    lastTest.Clear();
-                    lastTest.Dispose();
-                }
+                testContentContainer.Remove(lastTest.Parent);
+                lastTest.Dispose();
```
Comment (smoogipoo) [inline osu.Framework/Testing/TestBrowser.cs]: I have a feeling that this won't dispose `lastTest` if it wasn't added to the hierarchy.

Imagine the case where you click on a new test many times very quickly.

File: osu.Framework/Graphics/Drawable.cs [hunk 7]
```diff
@@ -135,22 +142,32 @@ private void dispose(bool isDisposing)
...
             {
                 Debug.Assert(loadTask == null);
                 loadState = LoadState.Loading;
-                loadTask = Task.Factory.StartNew(() => Load(target.Clock, target.Dependencies), TaskCreationOptions.LongRunning);
+                loadTask = Task.Factory.StartNew(() =>
+                {
+                    if (cancellationToken.IsCancellationRequested)
+                        return;
+
+                    if (isDisposed || target.isDisposed)
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Drawable.cs]: Redundant? The cancellation token should handle this case.

File: osu.Framework/Graphics/Drawable.cs [hunk 8]
```diff
@@ -135,22 +142,32 @@ private void dispose(bool isDisposing)
...
+                        return;
+
+                    Load(target.Clock, target.Dependencies);
```
Comment (smoogipoo) [inline osu.Framework/Graphics/Drawable.cs]: `Load()` is what we want cancelled. At the moment cancelling the load of a `CompositeDrawable` will potentially not cancel the load of its children.

**Comment by peppy** [comment]
all comments above moot; this is wip and not ready or intended for review.

File: osu.Framework/Platform/GameHost.cs (hunk 5 above)
Comment (peppy) [reply osu.Framework/Platform/GameHost.cs]: will cache it as a completely incorrect and inaccessible private type

File: osu.Framework/IO/Network/WebRequest.cs (hunk 4 above)
Comment (peppy) [reply osu.Framework/IO/Network/WebRequest.cs]: it gets disposed elsewhere, see rest of class

Comment (smoogipoo) [reply osu.Framework/IO/Network/WebRequest.cs]: Documenting IRL discussion: we actually don't even need the using in `internalPerform()`, but we do want the using in `PerformAsync()` as per my original comment.

**Comment by peppy** [comment]
Has been fulfilled via #1835

**Pull request closed without merging**